## 字串: 取得子字串

Python 字串編號圖示：

![Alt text](images/substring.png)


幾個子字串常見的寫法： 
* :冒號前的 INDEX 省略代表從 0 開始。
* :冒號後的 INDEX 省略代表直到最後一個字元。
*        str[INDEX] 取得 INDEX 的字元  
*        str[-INDEX] 取得倒數 INDEX 的字元  
*    str[START:END] 從 START 到 END-1 的子字串  
*    str[START:END:STEP] 從 START 到 END-1 每隔 STEP 的子字串  

In [23]:
s = 'Monty Python'
print("s[0]=",s[0]) # 字串 s 的第 0 個字元，python 使用 0 based 索引
print("s[6:12]=",s[6:12]) # 6~11 的子字串 6<=index<12
print("s[6:]=",s[6:]) # 6~字串尾端 的子字串
print("s[:6]=",s[:6]) # 0～5 子字串
print("s[:]=", s[:])
print("s[6:-2]=",s[6:-2]) # 6 代表倒數第 2位
print("s[-6:]=",s[-6:]) # 取得最後 6 位

print("s[0:6:2]=",s[0:6:2]) # 0~5 間隔 2 個字元
print("s[::2]=",s[::2]) # 從 0 開始間隔 2 個字元
print("s[::-1]=",s[::-1]) # 獲得一個倒反的字串

s[0]= M
s[6:12]= Python
s[6:]= Python
s[:6]= Monty 
s[:]= Monty Python
s[6:-2]= Pyth
s[-6:]= Python
s[0:6:2]= Mny
s[::2]= MnyPto
s[::-1]= nohtyP ytnoM


## 格式化輸出

### 方法 str.format() 的基本用法如下:



In [24]:
print('We are the {} who say "{}!"'.format('knights', 'Ni'))
print('We are the '+knights+' who say "'+Ni+'!"')
print('{0} and {1}'.format('spam', 'eggs'))
print('{1} and {0}'.format('spam', 'eggs'))
print('{1} and {0} and {1}'.format('spam', 'eggs'))

# 可以直接呼叫關鍵字參數
print('This {food} is {adjective}.'.format(food='eggs', adjective='absolutely horrible'))
# 位置參數和關鍵字參數可以隨意組合:
print('The story of {0}, {1}, and {other}.'.format('Bill', 'Manfred', other='Georg'))

We are the knights who say "Ni!"
spam and eggs
eggs and spam
eggs and spam and eggs
This eggs is absolutely horrible.
The story of Bill, Manfred, and Georg.


## Row String

\t\n\r 特殊符號通通無效


## python 3.6 以後新的 f-strings 寫法 (Literal String Interpolation)

In [5]:
import math 

name = "Tom"
age = 3
print(f"His name is {name}, he's {age} years old.")

se = 2
ss = f"The serial is {se:05d}"
print(ss)

PI = math.pi
print(f"Pi is {PI:.2f}")

#### 進制轉換
print(f'int: 31, hex: {31:x}, oct: {31:o}'
'int: 31, hex: 1f, oct: 37')

#### 支援生成式
print(f'Numbers from 1-10 are {[_ for _ in range(1, 11)]}')

#### 呼叫函數

def upper(s):
     return s.upper()

stock = 'tsmc'
close = 217.5
print(f'{stock} price: {close}')

print(f'{upper(stock)} price: {close}')

His name is Tom, he's 3 years old.
The serial is 00002
The serial is 00002
Pi is 3.14
int: 31, hex: 1f, oct: 37int: 31, hex: 1f, oct: 37
Numbers from 1-10 are [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
tsmc price: 217.5
TSMC price: 217.5


## 常見的格式化參數

<table class="code">
<tbody><tr>
<th> Number </th>
<th> Format </th>
<th> Output </th>
<th> Description </th>
</tr>
<tr>
<td> 3.1415926  </td>
<td> {:.2f}  </td>
<td> 3.14  </td>
<td> 2 decimal places </td>
</tr>
<tr>
<td> 3.1415926  </td>
<td> {:+.2f} </td>
<td> +3.14 </td>
<td> 2 decimal places with sign </td>
</tr>
<tr>
<td> -1         </td>
<td> {:+.2f} </td>
<td> -1.00 </td>
<td> 2 decimal places with sign </td>
</tr>
<tr>
<td> 2.71828    </td>
<td> {:.0f}  </td>
<td> 3   </td>
<td> No decimal places </td>
</tr>
<tr>
<td> 5          </td>
<td> {:0&gt;2d} </td>
<td> 05  </td>
<td> Pad number with zeros (left padding, width 2) </td>
</tr>
<tr>
<td> 5          </td>
<td> {:x&lt;4d} </td>
<td> 5xxx  </td>
<td> Pad number with x’s (right padding, width 4) </td>
</tr>
<tr>
<td> 10         </td>
<td> {:x&lt;4d} </td>
<td> 10xx  </td>
<td> Pad number with x’s (right padding, width 4) </td>
</tr>
<tr>
<td> 1000000    </td>
<td> {:,}    </td>
<td> 1,000,000  </td>
<td> Number format with comma separator </td>
</tr>
<tr>
<td> 0.25       </td>
<td> {:.2%}  </td>
<td> 25.00%  </td>
<td> Format percentage </td>
</tr>
<tr>
<td> 1000000000 </td>
<td> {:.2e}  </td>
<td> 1.00e+09 </td>
<td> Exponent notation </td>
</tr>
<tr>
<td> 13         </td>
<td> {:10d}  </td>
<td> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;13 </td>
<td> Right aligned (default, width 10) </td>
</tr>
<tr>
<td> 13         </td>
<td> {:&lt;10d} </td>
<td> 13 </td>
<td> Left aligned (width 10) </td>
</tr>
<tr>
<td> 13         </td>
<td> {:^10d} </td>
<td> &nbsp;&nbsp;&nbsp;&nbsp;13 </td>
<td> Center aligned (width 10) </td>
</tr>
</tbody></table>


①.當被修約的值爲5時，如果他前面的數爲偶數且被修約數的後面沒有數時則捨棄；

②.當被修約的值爲5時，如果他前面的數爲奇數時則進位；

③.當被修約的值爲5時，如果他前面的數爲偶數時且被修約數的後面還有數時，則進位。

寫的有點繁瑣，總之就一個概念“四捨六入五成雙”


In [22]:
print('固定寬度: {:7d}'.format(23000))
print('固定寬度不足補 0 : {:05d}'.format(9))
print('固定寬度不足補 0 : {:05d}'.format(29))
print('指定小數點位數為2 : {:.2f}'.format(3.14159))
print('指定小數點位數為0 : {:.0f}'.format(4.5))
print('四捨六入雙進位 : {:.0f}, {:.0f}, {:.0f}, {:.0f}'.format(1.5, 2.5, 3.5, 2.50001))

print('顯示為百分比 : {:.0%}'.format(0.45))

# 舊式 代位字元
print('The value of PI is approximately %5.3f.' % 3.14159)
print('The value of PI is approximately {:5.3f}.'.format(3.14159))


固定寬度:   23000
固定寬度不足補 0 : 00009
固定寬度不足補 0 : 00029
指定小數點位數為2 : 3.14
指定小數點位數為0 : 4
四捨六入雙進位 : 2, 2, 4, 3
顯示為百分比 : 45%
The value of PI is approximately 3.142.
The value of PI is approximately 3.142.


## 字串函數


In [2]:
s = "Hello, " + "Beauty... "*3 # 字串重複
print("s="+s)
print(s.find('B')) # 由左邊找起第一個找到的索引
print(s.replace('Beauty', 'Ugly')) # 取代某個字串
print(s.split()) # 字串分割 預設用 " " 分割
print(s.split('B')) # 字串分割 用 "," 分割
print(s.upper()) # 將字母變成大寫
print(s.lower()) # 將字母變成小寫
print(','.join("1 2 3 4".split()))
print(','.join(s.split())) # List 的每個元素中插入一個 ,
print(s.count('Beauty')) # 計算字串出現的次數
print(s.strip()) # 清除掉頭尾的不可見字元

print("strip: ",s.strip(',.-')) # 清除掉頭尾的字元

#將姓名中間兩個子替換成相同個數的 O
s = "王大同"
s = s.replace(s[1], "O")
print(s)

s = "村上春樹"
s = s.replace(s[1:-1], "O"*len(s[1:-1]))
print(s)

# 字元與整數互換。透過 ASCII 表。
# chr(97) ==》 A
# ord("A") ==》 97

s=Hello, Beauty... Beauty... Beauty... 
7
Hello, Ugly... Ugly... Ugly... 
['Hello,', 'Beauty...', 'Beauty...', 'Beauty...']
['Hello, ', 'eauty... ', 'eauty... ', 'eauty... ']
HELLO, BEAUTY... BEAUTY... BEAUTY... 
hello, beauty... beauty... beauty... 
1,2,3,4
Hello,,Beauty...,Beauty...,Beauty...
3
Hello, Beauty... Beauty... Beauty...
strip:  Hello, Beauty... Beauty... Beauty... 
王O同
村OO樹
